File: server/src/routes/member.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { authMiddleware } from '../middleware/auth';
import { requireCurrentOrgId, requireOrgRole, respondValidationError } from './_shared';
import {
  createMember,
  createMemberGroup,
  deleteMember,
  deleteMemberGroup,
  MemberMutationError,
  moveMemberToGroup,
  updateMember,
  updateMemberGroup,
} from '../features/member/mutations';
import {
  getMemberSummaryWithPackages,
  getMemberWithPackages,
  listMemberGroups,
  listMembers,
} from '../features/member/queries';
import { serializeMember } from '../features/member/serializer';

const router = Router();
router.use(authMiddleware);

const createMemberGroupSchema = z.object({
  name: z.string().min(1).max(40),
});

const updateMemberGroupSchema = z.object({
  name: z.string().min(1).max(40).optional(),
  sortOrder: z.number().int().min(0).optional(),
});

const moveMemberGroupSchema = z.object({
  memberGroupId: z.string().uuid().nullable(),
});

router.get('/groups', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;

    res.json(await listMemberGroups(orgId));
  } catch (err) {
    console.error('List member groups error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.post('/groups', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;
    if (!(await requireOrgRole(req, res, orgId, ['OWNER', 'MANAGER']))) return;

    const body = createMemberGroupSchema.parse(req.body);
    const group = await createMemberGroup({
      organizationId: orgId,
      name: body.name,
    });

    res.status(201).json(group);
  } catch (err) {
    if (respondValidationError(res, err)) return;
    console.error('Create member group error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.put('/groups/:id', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;
    if (!(await requireOrgRole(req, res, orgId, ['OWNER', 'MANAGER']))) return;

    const body = updateMemberGroupSchema.parse(req.body);
    const updated = await updateMemberGroup({
      organizationId: orgId,
      groupId: req.params.id as string,
      name: body.name,
      sortOrder: body.sortOrder,
    });
    res.json(updated);
  } catch (err) {
    if (err instanceof MemberMutationError && err.code === 'GROUP_NOT_FOUND') {
      res.status(404).json({ error: 'Group not found' });
      return;
    }
    if (respondValidationError(res, err)) return;
    console.error('Update member group error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.delete('/groups/:id', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;
    if (!(await requireOrgRole(req, res, orgId, ['OWNER', 'MANAGER']))) return;

    await deleteMemberGroup({
      organizationId: orgId,
      groupId: req.params.id as string,
    });

    res.json({ message: 'Group deleted' });
  } catch (err) {
    if (err instanceof MemberMutationError && err.code === 'GROUP_NOT_FOUND') {
      res.status(404).json({ error: 'Group not found' });
      return;
    }
    console.error('Delete member group error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── List Members ──────────────────────────────────────────
router.get('/', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;

    const { search, status } = req.query;
    const members = await listMembers({
      organizationId: orgId,
      search: typeof search === 'string' ? search : undefined,
      status: typeof status === 'string' ? status : undefined,
    });
    res.json(members.map(serializeMember));
  } catch (err) {
    console.error('List members error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── Get Member ────────────────────────────────────────────
router.get('/:id', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;

    const member = await getMemberWithPackages({
      organizationId: orgId,
      memberId: req.params.id as string,
    });
    if (!member) { res.status(404).json({ error: 'Member not found' }); return; }
    res.json(serializeMember(member));
  } catch (err) {
    console.error('Get member error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── Create Member ─────────────────────────────────────────
const createMemberSchema = z.object({
  name: z.string().min(1),
  phone: z.string().optional(),
  email: z.string().email().optional(),
  birthDate: z.string().optional(),
  gender: z.enum(['MALE', 'FEMALE', 'OTHER']).optional(),
  quickMemo: z.string().optional(),
  memo: z.string().optional(),
});

router.post('/', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;
    if (!(await requireOrgRole(req, res, orgId, ['OWNER', 'MANAGER']))) return;

    const body = createMemberSchema.parse(req.body);
    const member = await createMember({
      organizationId: orgId,
      ...body,
    });

    res.status(201).json(serializeMember(member));
  } catch (err) {
    if (respondValidationError(res, err)) return;
    console.error('Create member error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── Update Member ─────────────────────────────────────────
const updateMemberSchema = z.object({
  name: z.string().min(1).optional(),
  phone: z.string().optional(),
  email: z.string().email().optional(),
  birthDate: z.string().optional(),
  gender: z.enum(['MALE', 'FEMALE', 'OTHER']).optional(),
  quickMemo: z.string().optional(),
  memo: z.string().optional(),
  status: z.enum(['ACTIVE', 'INACTIVE', 'ARCHIVED']).optional(),
});

router.put('/:id', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;
    if (!(await requireOrgRole(req, res, orgId, ['OWNER', 'MANAGER']))) return;

    const body = updateMemberSchema.parse(req.body);
    const updated = await updateMember({
      organizationId: orgId,
      memberId: req.params.id as string,
      ...body,
    });
    if (!updated) {
      res.json(updated);
      return;
    }
    res.json(serializeMember(updated));
  } catch (err) {
    if (err instanceof MemberMutationError && err.code === 'MEMBER_NOT_FOUND') {
      res.status(404).json({ error: 'Member not found' });
      return;
    }
    if (respondValidationError(res, err)) return;
    console.error('Update member error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.patch('/:id/group', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;
    if (!(await requireOrgRole(req, res, orgId, ['OWNER', 'MANAGER']))) return;

    const body = moveMemberGroupSchema.parse(req.body);
    const member = await moveMemberToGroup({
      organizationId: orgId,
      memberId: req.params.id as string,
      memberGroupId: body.memberGroupId,
    });
    if (!member) {
      res.status(404).json({ error: 'Member not found' });
      return;
    }
    res.json(serializeMember(member));
  } catch (err) {
    if (err instanceof MemberMutationError) {
      if (err.code === 'GROUP_NOT_FOUND') {
        res.status(404).json({ error: 'Group not found' });
        return;
      }
      if (err.code === 'MEMBER_NOT_FOUND') {
        res.status(404).json({ error: 'Member not found' });
        return;
      }
    }
    if (respondValidationError(res, err)) return;
    console.error('Move member group error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── Delete Member ─────────────────────────────────────────
router.delete('/:id', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;
    if (!(await requireOrgRole(req, res, orgId, ['OWNER', 'MANAGER']))) return;

    await deleteMember({
      organizationId: orgId,
      memberId: req.params.id as string,
    });
    res.json({ message: 'Member deleted' });
  } catch (err) {
    if (err instanceof MemberMutationError && err.code === 'MEMBER_NOT_FOUND') {
      res.status(404).json({ error: 'Member not found' });
      return;
    }
    console.error('Delete member error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: app/lib/models/member.dart
class MemberGroup {
  final String id;
  final String organizationId;
  final String name;
  final int sortOrder;

  const MemberGroup({
    required this.id,
    required this.organizationId,
    required this.name,
    required this.sortOrder,
  });

  factory MemberGroup.fromJson(Map<String, dynamic> json) {
    return MemberGroup(
      id: json['id'] as String,
      organizationId: json['organizationId'] as String,
      name: json['name'] as String,
      sortOrder: json['sortOrder'] as int? ?? 0,
    );
  }
}

class Member {
  final String id;
  final String organizationId;
  final String? memberGroupId;
  final String? memberGroupName;
  final int memberGroupSortOrder;
  final String name;
  final String? phone;
  final String? email;
  final DateTime? birthDate;
  final String? gender;
  final String? quickMemo;
  final String? memo;
  final String status;
  final String packageStatus;
  final String packageStatusLabel;
  final bool hasMemberAccount;
  final String memberSourceLabel;
  final String memberAccessLabel;
  final DateTime createdAt;

  const Member({
    required this.id,
    required this.organizationId,
    this.memberGroupId,
    this.memberGroupName,
    this.memberGroupSortOrder = 0,
    required this.name,
    this.phone,
    this.email,
    this.birthDate,
    this.gender,
    this.quickMemo,
    this.memo,
    required this.status,
    required this.packageStatus,
    required this.packageStatusLabel,
    required this.hasMemberAccount,
    required this.memberSourceLabel,
    required this.memberAccessLabel,
    required this.createdAt,
  });

  factory Member.fromJson(Map<String, dynamic> json) {
    return Member(
      id: json['id'] as String,
      organizationId: json['organizationId'] as String,
      memberGroupId: json['memberGroupId'] as String?,
      memberGroupName:
          (json['memberGroup'] as Map<String, dynamic>?)?['name'] as String?,
      memberGroupSortOrder:
          (json['memberGroup'] as Map<String, dynamic>?)?['sortOrder']
              as int? ??
          0,
      name: json['name'] as String,
      phone: json['phone'] as String?,
      email: json['email'] as String?,
      birthDate: json['birthDate'] != null
          ? DateTime.parse(json['birthDate'] as String)
          : null,
      gender: json['gender'] as String?,
      quickMemo: json['quickMemo'] as String?,
      memo: json['memo'] as String?,
      status: json['status'] as String? ?? 'ACTIVE',
      packageStatus: json['packageStatus'] as String? ?? 'GENERAL_MEMBER',
      packageStatusLabel: json['packageStatusLabel'] as String? ?? '일반 회원',
      hasMemberAccount: json['hasMemberAccount'] as bool? ?? false,
      memberSourceLabel: json['memberSourceLabel'] as String? ?? '관리자 등록 회원',
      memberAccessLabel: json['memberAccessLabel'] as String? ?? '채팅 미연동',
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'memberGroupId': memberGroupId,
    'name': name,
    'phone': phone,
    'email': email,
    'birthDate': birthDate?.toIso8601String(),
    'gender': gender,
    'quickMemo': quickMemo,
    'memo': memo,
    'status': status,
    'packageStatus': packageStatus,
    'packageStatusLabel': packageStatusLabel,
    'hasMemberAccount': hasMemberAccount,
    'memberSourceLabel': memberSourceLabel,
    'memberAccessLabel': memberAccessLabel,
  };

  Member copyWith({
    String? name,
    String? phone,
    String? email,
    DateTime? birthDate,
    String? gender,
    String? quickMemo,
    String? memo,
    String? status,
    String? packageStatus,
    String? packageStatusLabel,
    bool? hasMemberAccount,
    String? memberSourceLabel,
    String? memberAccessLabel,
  }) {
    return Member(
      id: id,
      organizationId: organizationId,
      memberGroupId: memberGroupId,
      memberGroupName: memberGroupName,
      memberGroupSortOrder: memberGroupSortOrder,
      name: name ?? this.name,
      phone: phone ?? this.phone,
      email: email ?? this.email,
      birthDate: birthDate ?? this.birthDate,
      gender: gender ?? this.gender,
      quickMemo: quickMemo ?? this.quickMemo,
      memo: memo ?? this.memo,
      status: status ?? this.status,
      packageStatus: packageStatus ?? this.packageStatus,
      packageStatusLabel: packageStatusLabel ?? this.packageStatusLabel,
      hasMemberAccount: hasMemberAccount ?? this.hasMemberAccount,
      memberSourceLabel: memberSourceLabel ?? this.memberSourceLabel,
      memberAccessLabel: memberAccessLabel ?? this.memberAccessLabel,
      createdAt: createdAt,
    );
  }
}
